### PR TITLE
[mlir] Do not merge blocks during canonicalization by default

### PR DESCRIPTION
**Pull request by joker-eph**
This is a heavy process, and it can trigger a massive explosion in adding block arguments. While potentially reducing the code size, the resulting merged blocks with arguments are hiding some of the def-use chain and can even hinder some further analyses/optimizations: a merge block does not have it's own path-sensitive context, instead the context is merged from all the predecessors.

Previous behavior can be restored by passing:

  {test-convergence region-simplify=aggressive}

to the canonicalize pass.

### CODE REVIEW

**Review by Mogball — APPROVED** [review]
yeah I noticed the explosion behaviour in https://github.com/llvm/llvm-project/issues/63230

File: mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h [hunk 1]
```diff
@@ -29,6 +29,16 @@ enum class GreedyRewriteStrictness {
   ExistingOps
 };
 
+enum class GreedySimplifyRegionLevel {
```
Comment (bondhugula) [inline mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h:32]: Doc comment for this.

File: mlir/lib/Transforms/Utils/RegionUtils.cpp [hunk 2]
```diff
@@ -827,11 +827,13 @@ static LogicalResult mergeIdenticalBlocks(RewriterBase &rewriter,
...
-                                    MutableArrayRef<Region> regions) {
+                                    MutableArrayRef<Region> regions,
+                                    bool mergeBlocks) {
   bool eliminatedBlocks = succeeded(eraseUnreachableBlocks(rewriter, regions));
   bool eliminatedOpsOrArgs = succeeded(runRegionDCE(rewriter, regions));
-  bool mergedIdenticalBlocks =
-      succeeded(mergeIdenticalBlocks(rewriter, regions));
+  bool mergedIdenticalBlocks = false;
+  if (mergeBlocks)
+    mergedIdenticalBlocks = succeeded(mergeIdenticalBlocks(rewriter, regions));
```
Comment (bondhugula) [inline mlir/lib/Transforms/Utils/RegionUtils.cpp:836]: Can be written in one line:
```
bool mergedIdenticalBlocks = mergeBlocks && succeeded(...);
```

File: mlir/test/Transforms/canonicalize-block-merge.mlir [hunk 3]
```diff
@@ -275,3 +275,19 @@ func.func @mismatch_dominance() -> i32 {
 ^bb4(%3: i32):
   return %3 : i32
 }
+
```
Comment (bondhugula) [inline mlir/test/Transforms/canonicalize-block-merge.mlir:278]: Drop extra line.

**Review by bondhugula — APPROVED** [review]
This looks good to me. Minor comments.

**Review by giuseros — COMMENTED** [review]
Thanks for the PR, I had one ready to introduce a `numBlocksArgs` threshold. I am not sure which one is better, but my idea was that reducing the number of blocks, up to a point, was something useful (especially on GPUs where branches are so bad)

File: mlir/lib/Transforms/Utils/RegionUtils.cpp (hunk 2 above)
Comment (joker-eph) [reply mlir/lib/Transforms/Utils/RegionUtils.cpp:836]: I don't really like the short circuit with side effects: this would obscure the control flow IMO.

**Comment by joker-eph** [comment]
> I had one ready to introduce a numBlocksArgs threshold. I am not sure which one is better, but my idea was that reducing the number of blocks, up to a point, was something useful (especially on GPUs where branches are so bad)

What is the problem you're seeing with the number of block arguments? (in terms of code generation they are lowered to "nothing", basically just a register constraint, worst case a register copy)

**Comment by giuseros** [comment]
> What is the problem you're seeing with the number of block arguments? (in terms of code generation they are lowered to "nothing", basically just a register constraint, worst case a register copy)

The problem is that blocks with >32k arguments are very slow to manage and hang the compiler (with 32k it takes few minutes, but the original IR I had did never stop).

**Comment by joker-eph** [comment]
> The problem is that blocks with >32k arguments are very slow to manage and hang the compiler (with 32k it takes few minutes, but the original IR I had did never stop compiling).

It'd be nice to check if we have some obvious issues with our block argument scaling! Do you have some IR reproducer you could share?

**Comment by giuseros** [comment]
> It'd be nice to check if we have some obvious issues with our block argument scaling! Do you have some IR reproducer you could share?

Yes, I am attaching the IRs I used to investigate this issue to this comment. 

If you do:
```
$ ./bin/mlir-opt --canonicalize disable_simplify.mlir
```

The compiler will hang for about 4 seconds and then complete. But if you run:
```
$ ./bin/mlir-opt --canonicalize disable_simplify_hang.mlir
```
The compiler will hang for a long time (I never got it to finish). If you pass `region-simplify=false` it will finish almost immediately. 

[disable_simplify.zip](https://github.com/user-attachments/files/15801638/disable_simplify.zip)

File: mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h [hunk 4]
```diff
@@ -45,7 +55,8 @@ class GreedyRewriteConfig {
   /// patterns.
   ///
   /// Note: Only applicable when simplifying entire regions.
-  bool enableRegionSimplification = true;
+  GreedySimplifyRegionLevel enableRegionSimplification =
+      GreedySimplifyRegionLevel::Aggressive;
```
Comment (giuseros) [inline mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h:59]: Sorry, I didn't spot this, but I think this should be set as `Normal`. Having `Aggressive` by default will make the compiler hang for large merges.

Comment (joker-eph) [reply mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h:59]: This is preserving the behavior of the GreedyPatternRewriter.
This is not the default for the canonicalize pass.

Comment (giuseros) [reply mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h:59]: The point is that the initial bug was seen in the rewriter (canonicalization was the way I reproduced it). Maybe we could set a threshold for the number of block arguments to merge (like I was proposing above)?

Comment (joker-eph) [reply mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h:59]: > The point is that the initial bug was seen in the rewriter 

My point is that clients of the rewriter for purposes other than canonicalization should decide what they want. I'm rather confident in changing this default for the purpose of canonicalization, I don't think I need to change other usage of the greedy driver right now.

>  Maybe we could set a threshold for the number of block arguments to merge (like I was proposing above)?

That seems like a clunk to me to be honest. I would rather investigate the issue first.

Comment (giuseros) [reply mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h:59]: Ok, that's fine. I will try to work on the core issue, and meanwhile we can configure the rewriter to do a `Normal` region simplification.

Comment (giuseros) [reply mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h:59]: Thanks @joker-eph

**Comment by giuseros** [comment]
Hi @joker-eph , should we merge this in?

**Comment by joker-eph** [comment]
I was hoping to hear from @jpienaar actually :)

**Review by jpienaar — APPROVED** [review]
This makes sense to me. The aggressive configuration could get a param later (if needed) but almost feels like it could be a separate pass too. Not a blocker here though.

This enables the current state while also allowing folks (probably majority but we'll probably only find out in post if I'm wrong :-)) to avoid the cost.

**Pull request merged**